spring-cloud-kubernetes-commons/src/main/java/org/springframework/cloud/kubernetes/commons/config/configdata/KubernetesConfigDataLocationResolver.java

```diff
@@ -0,0 +1,134 @@
+		this.log = factory.getLog(KubernetesConfigDataLocationResolver.class);
+	}
+
+	protected final String getPrefix() {
+		return "kubernetes:";
+	}
+
+	@Override
+	public final int getOrder() {
+		return -1;
+	}
+
+	@Override
+	public final boolean isResolvable(ConfigDataLocationResolverContext context, ConfigDataLocation location) {
+		return location.hasPrefix(getPrefix())
+				&& (KUBERNETES.isEnforced(context.getBinder()) || KUBERNETES.isDetected(new StandardEnvironment()));
+	}
+
+	@Override
+	public final List<KubernetesConfigDataResource> resolve(ConfigDataLocationResolverContext context,
+			ConfigDataLocation location)
+			throws ConfigDataLocationNotFoundException, ConfigDataResourceNotFoundException {
+		return Collections.emptyList();
+	}
+
+	@Override
+	public final List<KubernetesConfigDataResource> resolveProfileSpecific(
+			ConfigDataLocationResolverContext resolverContext, ConfigDataLocation location, Profiles profiles)
+			throws ConfigDataLocationNotFoundException {
+
+		ConfigDataProperties properties = ConfigDataProperties.of(resolverContext).register();
+
+		Map<String, Object> kubernetesConfigData = new HashMap<>();
+		kubernetesConfigData.put("spring.cloud.kubernetes.client.namespace", properties.clientProperties().namespace());
+		String applicationName = resolverContext.getBinder().bind("spring.application.name", String.class).orElse(null);
+		if (applicationName != null) {
+			kubernetesConfigData.put("spring.application.name", applicationName);
+		}
+
+		Environment environment = environment(kubernetesConfigData, profiles);
+		KubernetesNamespaceProvider namespaceProvider = kubernetesNamespaceProvider(environment);
+		registerBeans(resolverContext, location, profiles, properties, namespaceProvider);
+
+		KubernetesConfigDataResource resource = new KubernetesConfigDataResource(properties.clientProperties(),
+				properties.configMapProperties(), properties.secretsProperties(), location.isOptional(), profiles,
+				environment);
+
+		return List.of(resource);
+	}
+
+	protected abstract void registerBeans(ConfigDataLocationResolverContext resolverContext,
+			ConfigDataLocation location, Profiles profiles, ConfigDataProperties properties,
+			KubernetesNamespaceProvider namespaceProvider);
+
+	protected final boolean isRetryEnabledForConfigMap(ConfigMapConfigProperties configMapProperties) {
+		return RETRY_IS_PRESENT && configMapProperties != null && configMapProperties.retry().enabled()
+				&& configMapProperties.failFast();
+	}
+
+	protected final boolean isRetryEnabledForSecrets(SecretsConfigProperties secretsProperties) {
+		return RETRY_IS_PRESENT && secretsProperties != null && secretsProperties.retry().enabled()
+				&& secretsProperties.failFast();
+	}
+
+	private KubernetesNamespaceProvider kubernetesNamespaceProvider(Environment environment) {
+		return new KubernetesNamespaceProvider(environment);
+	}
+
+	private Environment environment(Map<String, Object> kubernetesConfigData, Profiles profiles) {
```

Comment:
extract to a separate method, making `resolveProfileSpecific` even easier to read